OIDC provider

sorenisanerd opened the pull request:
This adds support for a Github Actions compatible OIDC provider.

For more info, see:
   https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect

It depends on a few changes in act and act_runner:
https://gitea.com/gitea/act_runner/pulls/272
https://gitea.com/gitea/act/pulls/73

sorenisanerd committed:
Refactor token signing method in OAuth2 service

sorenisanerd commented:
Still work-in-progress. I have a few more things I want to test before it's ready.

lunny commented on go.mod:
```diff
@@ -300,7 +300,7 @@ replace github.com/hashicorp/go-version => github.com/6543/go-version v1.3.1
 replace github.com/shurcooL/vfsgen => github.com/lunny/vfsgen v0.0.0-20220105142115-2c99e1ffdfa0
 
-replace github.com/nektos/act => gitea.com/gitea/act v0.243.4
+replace github.com/nektos/act => gitea.com/sorenisanerd/act v0.246.1-0.20230704003127-0ffa64475fd8
```
Please send PR to gitea.com/gitea/act first.

sorenisanerd replied:
I did. https://gitea.com/gitea/act/pulls/73

sorenisanerd commented:
I think this is ready for review, but obviously it's useless without the act and act_runner changes.

wolfogre commented:
@sorenisanerd Thank you for your work, I have to say you did a great job, good idea, good code.

However, this featrue has been blocked some trivial things.

- Where to keep the code for parsing permissions? Gitea, the forked act, the upstream act, or using actionlint indead?
- The `permissions` field is only only for OIDC, but the permissions granted to the `GITHUB_TOKEN/GITEA_TOKEN` (probably more often), personally, I think the latter should be supported first.
- Is it possible for contributors to send a PR that updates the `permissions` field and runs a malicious script?

You know, once permissions are involved, we need to be checked more carefully. Please give us some time to discuss and decide.

sorenisanerd commented:
Thank you for your kind comments :)

I'm very open to porting this to gitea or actionlinter or wherever. I don't have any strong opinions on it.

I agree it would have made more sense to first use permissions to adjust the capabilities of the GitHub token. I happened to need this oidc-provider functionality, so that's what I focused on. I included methods for merging permissions as well as definitions matching GitHub's defaults so I or someone else could add that later.

Great point about how permissions are affected when permissions apply for pull requests. Is that supposed to be safeguarded by the need to approve workflow runs for new contributors? I honestly don't know what GitHub does.

sorenisanerd commented:
I've pushed changes to the act and act_runner PR's on gitea.com. The resulting diffs are very modest. The rest of the functionality has been pulled into Gitea here.

Is this approach more agreeable? I'll add some tests once we're happy with the overall approach.

sorenisanerd committed:
feat: Add basic permissions support for actions

This depends on a few changes in act and act_runner:

  https://gitea.com/gitea/act_runner/pulls/272
  https://gitea.com/gitea/act/pulls/73

sorenisanerd committed:
Extract {base_,}{sha,ref} and EventName logic

sorenisanerd committed:
feat: Add OIDC provider for actions

sorenisanerd commented:
I rebased this and reworked it into a set of more reviewable commits. If you like, I can submit them as individual PR's, but each one depends on the previous ones, which makes it a little awkward to manage as separate PR's.

sorenisanerd commented:
Nothing new, just rebased and folded https://github.com/go-gitea/gitea/commit/ded82c30abd033695d7a6edae9a1295f792063a0 (the formatting changes based on the linting errors) into the relevant patches.

lunny commented on go.mod:
```diff
@@ -303,7 +303,7 @@ replace github.com/hashicorp/go-version => github.com/6543/go-version v1.3.1
 replace github.com/shurcooL/vfsgen => github.com/lunny/vfsgen v0.0.0-20220105142115-2c99e1ffdfa0
 
-replace github.com/nektos/act => gitea.com/gitea/act v0.243.4
+replace github.com/nektos/act => gitea.com/sorenisanerd/act v0.246.2-0.20230806181409-a9e947b70bf6
```
Maybe a update on gitea/act side is better.

sorenisanerd replied:
As mentioned in the PR description, there's a PR against gitea/act at https://gitea.com/gitea/act/pulls/73 since several months ago.

lunny commented on models/migrations/v1_21/v276.go:
```diff
@@ -0,0 +1,18 @@
+
+func AddPermissions(x *xorm.Engine) error {
+	type ActionRunJob struct {
+		Permissions actions_model.Permissions `xorm:"JSON TEXT"`
```
We need to copy the Permissions struct here because in future that struct maybe changed.

sorenisanerd replied:
Excellent point. I'll make that change.

sorenisanerd replied:
Wait, does it even matter, since it's stored as JSON? Should I just put `interface{}` instead of `actions_model.Permissions`?

lunny commented on routers/api/v1/oidc.go:
```diff
@@ -0,0 +1,154 @@
+// Copyright 2016 The Gitea Authors. All rights reserved.
```
2023

sorenisanerd replied:
lol. Right.

lunny commented on services/auth/oauth2.go:
```diff
@@ -103,6 +103,7 @@ func (o *OAuth2) userIDFromToken(ctx context.Context, tokenSHA string, store Dat
 
 				store.GetData()["IsActionsToken"] = true
 				store.GetData()["ActionsTaskID"] = task.ID
+				store.GetData()["ActionsTask"] = task
```
We can load the task from database when we need it but not store it in context?

sorenisanerd replied:
I honestly don't remember why I did it this way. It's been a while. Let me take another look.

sorenisanerd committed:
Copy Permissions struct actions_model

Referencing it could cause inconsistencies if it changes later.

sorenisanerd committed:
*looks at a calendar*

sorenisanerd committed:
Merge remote-tracking branch 'origin/main' into oidc-provider

sorenisanerd commented:
I took the liberty of renaming my migration to just `permissions.go`. At this point, I must have had to rename it 5-7 times locally and it's getting kinda annoying. Once this gets closer to being approved, I'll try to pick a number for it.

sorenisanerd commented:
I'm also working on a more complete permissions implementation that should probably be merged before this one, so I'm setting this one back to WIP. I expect the permissions patch to be ready some time this week.

punmechanic commented:
Hello, are there any updates on this? is there something a contributor can help out with? Would love to see this move along

NorNorLukas commented:
I am also very interested om seeing this moving along. Any updates on this?

scubbo commented:
Also chiming in support of this - it would be great for Gitea Actions to be able to authenticate to Vault via JWT.

sorenisanerd commented:
It was taking too much effort to merge/rebase it while it was waiting for
review, so I lost my motivation. I can't make promises about when I'll
revive it. If someone else wants to take it over, be my guest.

On Mon, Mar 17, 2025 at 9:05 PM Jack Jackson ***@***.***>
wrote:

> Also chiming in support of this - it would be great for Gitea Actions to
> be able to authenticate to Vault via JWT.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/go-gitea/gitea/pull/25664#issuecomment-2731558657>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABHCWU2YHQVOXC7UHOW3C32U6LQHAVCNFSM6AAAAABRF6YIUOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMZRGU2TQNRVG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: scubbo]*scubbo* left a comment (go-gitea/gitea#25664)
> <https://github.com/go-gitea/gitea/pull/25664#issuecomment-2731558657>
>
> Also chiming in support of this - it would be great for Gitea Actions to
> be able to authenticate to Vault via JWT.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/go-gitea/gitea/pull/25664#issuecomment-2731558657>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABHCWU2YHQVOXC7UHOW3C32U6LQHAVCNFSM6AAAAABRF6YIUOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMZRGU2TQNRVG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

scubbo commented:
Understandable - thank you for your effort so far!

I haven't made any contributions to the `gitea` codebase, so I'd mostly just be replicating what you've done and trying my best to understand it as I go - but I'm certainly willing to try!

gnanakeethan commented:
It seems a separate PR was merged https://gitea.com/gitea/act/pulls/133

Is there any relationship between these two?

lunny commented:
Replaced by #36988
